File: python/sol/monitor.py
# -*- coding: utf-8 -*-
"""
monitor.py
"""

import sys
import datetime

def log1(fp, msg):
    log1_(fp,         msg)
    log1_(sys.stdout, msg)

def log2(fp,
        GPU, Parm, Nx, Ny, Nz, NN,
        iMaterial, iGeometry, iFeed, iPoint, Freq1, Freq2,
        iPmlEx, iPmlEy, iPmlEz, iPmlHx, iPmlHy, iPmlHz):
    log2_(fp,
        GPU, Parm, Nx, Ny, Nz, NN,
        iMaterial, iGeometry, iFeed, iPoint, Freq1, Freq2,
        iPmlEx, iPmlEy, iPmlEz, iPmlHx, iPmlHy, iPmlHz)
    log2_(sys.stdout,
        GPU, Parm, Nx, Ny, Nz, NN,
        iMaterial, iGeometry, iFeed, iPoint, Freq1, Freq2,
        iPmlEx, iPmlEy, iPmlEz, iPmlHx, iPmlHy, iPmlHz)

def log3(fp, fn_log, fn_out):
    log3_(fp,         fn_log, fn_out)
    log3_(sys.stdout, fn_log, fn_out)

def log4(fp, cpu):
    log4_(fp,         cpu) 
    log4_(sys.stdout, cpu)

# (private) message
def log1_(fp, msg):
    fp.write('%s\n' % msg)
    fp.flush()

# (private) condition
def log2_(fp,
    GPU, Parm, Nx, Ny, Nz, NN,
    iMaterial, iGeometry, iFeed, iPoint, Freq1, Freq2,
    iPmlEx, iPmlEy, iPmlEz, iPmlHx, iPmlHy, iPmlHz):

    cpu_mem, gpu_mem = _memory_size(Parm, Nx, Ny, Nz, Freq2, NN,
        iPmlEx, iPmlEy, iPmlEz, iPmlHx, iPmlHy, iPmlHz)
    out = _output_size(NN, Freq2)

    fp.write("%s\n" % datetime.datetime.now().ctime())
    fp.write("Title = %s\n" % Parm['title'])
    fp.write("Source = %s\n" % ("feed" if Parm['source'] == 0 else "plane wave"))
    fp.write("Cells = %d x %d x %d = %d\n" % (Nx, Ny, Nz, Nx * Ny * Nz))
    fp.write("No. of Materials  = %d\n" % len(iMaterial))
    fp.write("No. of Geometries = %d\n" % iGeometry.shape[0])
    if Parm['source'] == 0:
        fp.write("No. of Feeds      = %d\n" % iFeed.shape[0])
    fp.write("No. of Points     = %d\n" % iPoint.shape[0])
    fp.write("No. of Freq.s (1) = %d\n" % len(Freq1))
    fp.write("No. of Freq.s (2) = %d\n" % len(Freq2))
    fp.write("CPU Memory size   = %d [MB]\n" % cpu_mem)
    if GPU:
        fp.write("GPU Memory size   = %d [MB]\n" % gpu_mem)
    fp.write("Output filesize   = %d [MB]\n" % out)
    if   Parm['abc'][0] == 0:
        fp.write("ABC = Mur-1st\n")
    elif Parm['abc'][0] == 1:
        fp.write("ABC = PML (L=%d, M=%.2f, R0=%.2e)\n" % (Parm['abc'][1], Parm['abc'][2], Parm['abc'][3]) )
    if (sum(Parm['pbc']) > 0):
        fp.write("PBC : %s%s%s\n" % (("X" if Parm['pbc'][0] == 1 else ""),
                                     ("Y" if Parm['pbc'][1] == 1 else ""),
                                     ("Z" if Parm['pbc'][2] == 1 else "")))
    fp.write("Dt[sec] = %.4e, Tw[sec] = %.4e, Tw/Dt = %.3f\n" % (Parm['dt'], Parm['tw'], Parm['tw'] / Parm['dt']))
    fp.write("Iterations = %d, Convergence = %.3e\n" % (Parm['solver'][0], Parm['solver'][2]))
    fp.write("=== iteration start ===\n")
    fp.write("   step   <E>      <H>\n")
    fp.flush()

# (private) output files
def log3_(fp, fn_log, fn_out):
    fp.write("%s\n" % "=== output files ===")
    fp.write("%s, %s\n" % (fn_log, fn_out))
    fp.flush()

# (private) cpu time
def log4_(fp, cpu):
    fp.write("=== cpu time [sec] ===\n")
    fp.write("  part-1 : %11.3f\n" % (cpu[1] - cpu[0]))
    fp.write("  part-2 : %11.3f\n" % (cpu[2] - cpu[1]))
    fp.write("  part-3 : %11.3f\n" % (cpu[3] - cpu[2]))
    fp.write("  part-4 : %11.3f\n" % (cpu[4] - cpu[3]))
    fp.write("  --------------------\n")
    fp.write("  total  : %11.3f\n" % (cpu[4] - cpu[0]))
    fp.write("=== normal end ===\n")
    fp.write("%s\n" % datetime.datetime.now().ctime())
    fp.flush()

# (private) メモリーサイズ [MB]
def _memory_size(Parm, Nx, Ny, Nz, Freq2, NN,
    iPmlEx, iPmlEy, iPmlEz, iPmlHx, iPmlHy, iPmlHz):

    f_dsize = 4 if Parm['f_dtype'] == 'f4' else 8
    i_dsize = 1 if Parm['i_dtype'] == 'u1' else 4

    mem  = NN * 6 * f_dsize         # Ex, Ey, Ez, Hx, Hy, Hz
    mem += NN * 6 * i_dsize         # iEx, iEy, iEz, iHx, iHy, iHz

    if Parm['vector']:
        mem += NN * 12 * f_dsize    # K1Ex, K2Ex, ... K1Hx, K2Hx, ...

    # PML
    if Parm['abc'][0] == 1:
        num = iPmlEx.shape[0] + iPmlEy.shape[0] + iPmlEz.shape[0] \
            + iPmlHx.shape[0] + iPmlHy.shape[0] + iPmlHz.shape[0]
        mem += num * ((4 * 4) + (2 * 6 * f_dsize))  # i, j, k, m, Exy, Exz, ...

    cpu_mem = mem + len(Freq2) * Nx * Ny * Nz * 6 * 8  # 'c8': cEx, cEy, cEz, cHx, cHy, cHz

    gpu_mem =     mem / (1024 * 1024) + 1
    cpu_mem = cpu_mem / (1024 * 1024) + 1

    return cpu_mem, gpu_mem

# (private) 出力ファイルサイズ [MB]
def _output_size(NN, Freq2):

    mem = 0
    mem += len(Freq2) * NN * 6 * 8  # 'c8': cEx, cEy, cEz, cHx, cHy, cHz

    return mem // (1024 * 1024) + 1

File: sol_sycl_dm/monitor.c
/*
monitor.c
*/

#include "ofd.h"

static void memory_size(int *, int *);
static int output_size(void);


// title or message
static void monitor1_(FILE *fp, const char msg[])
{
	fprintf(fp, "%s\n", msg);
	fflush(fp);
}


// condition
static void monitor2_(FILE *fp, int gpu, int commsize)
{
	int cpu_mem, gpu_mem;
	memory_size(&cpu_mem, &gpu_mem);

	time_t now;
	time(&now);

	fprintf(fp, "%s", ctime(&now));
	fprintf(fp, "Title = %s\n", Title);
	fprintf(fp, "Source = %s\n", (NFeed ? "feed" : "plane wave"));
	fprintf(fp, "Cells = %d x %d x %d = %lld\n", Nx, Ny, Nz, ((long long int)Nx * Ny * Nz));
	fprintf(fp, "No. of Materials   = %zd\n", NMaterial);
	fprintf(fp, "No. of Geometries  = %zd\n", NGeometry);
	if (NFeed) fprintf(fp, "No. of Feeds       = %d\n", NFeed);
	fprintf(fp, "No. of Points      = %d\n", NPoint);
	fprintf(fp, "No. of Freq.s (1)  = %d\n", NFreq1);
	fprintf(fp, "No. of Freq.s (2)  = %d\n", NFreq2);
	fprintf(fp, "CPU Memory size    = %d [MB]\n", cpu_mem);
	if (gpu) fprintf(fp, "GPU Memory %s = %d [MB]\n", (commsize > 1 ? "/ proc." : "size   "), gpu_mem);
	fprintf(fp, "Output filesize    = %d [MB]\n", output_size());
	if (iABC == 0) fprintf(fp, "ABC = Mur-1st\n");
	if (iABC == 1) fprintf(fp, "ABC = PML (L=%d, M=%.2f, R0=%.2e)\n", cPML.l, cPML.m, cPML.r0);
	if (PBCx || PBCy || PBCz) fprintf(fp, "PBC : %s%s%s\n", (PBCx ? "X" : ""), (PBCy ? "Y" : ""), (PBCz ? "Z" : ""));
	fprintf(fp, "Dt[sec] = %.4e, Tw[sec] = %.4e, Tw/Dt = %.3f\n", Dt, Tw, Tw / Dt);
	fprintf(fp, "Iterations = %d, Convergence = %.3e\n", Solver.maxiter, Solver.converg);
	fprintf(fp, "=== iteration start ===\n");
	fprintf(fp, "   step   <E>      <H>\n");

	fflush(fp);
}


// output files
static void monitor3_(FILE *fp, const char fn_log[], const char fn_out[], const char fn_feed[], const char fn_point[])
{
	fprintf(fp, "=== output files ===\n");
	fprintf(fp, "%s, %s", fn_log, fn_out);
	if (NFeed && NFreq1) {
		fprintf(fp, ", %s", fn_feed);
	}
	if (NPoint && NFreq1) {
		fprintf(fp, ", %s", fn_point);
	}
	fprintf(fp, "\n");
	fflush(fp);
}


// cpu time
static void monitor4_(FILE *fp, const double cpu[])
{
	time_t now;
	time(&now);

	fprintf(fp, "%s\n", "=== cpu time [sec] ===");
	fprintf(fp, "  part-1 : %11.3f\n", cpu[1] - cpu[0]);
	fprintf(fp, "  part-2 : %11.3f\n", cpu[2] - cpu[1]);
	fprintf(fp, "  part-3 : %11.3f\n", cpu[3] - cpu[2]);
	fprintf(fp, "  part-4 : %11.3f\n", cpu[4] - cpu[3]);
	fprintf(fp, "  %s\n", "--------------------");
	fprintf(fp, "  total  : %11.3f\n", cpu[4] - cpu[0]);
	fprintf(fp, "%s\n", "=== normal end ===");
	fprintf(fp, "%s", ctime(&now));

	fflush(fp);
}


void monitor1(FILE *fp, const char msg[])
{
	monitor1_(fp,     msg);
	monitor1_(stdout, msg);
}


void monitor2(FILE *fp, int gpu, int commsize)
{
	monitor2_(fp,     gpu, commsize);
	monitor2_(stdout, gpu, commsize);
}


void monitor3(FILE *fp, const char fn_log[], const char fn_out[], const char fn_feed[], const char fn_point[])
{
	monitor3_(fp,     fn_log, fn_out, fn_feed, fn_point);
	monitor3_(stdout, fn_log, fn_out, fn_feed, fn_point);
}


void monitor4(FILE *fp, const double cpu[])
{
	monitor4_(fp,     cpu);
	monitor4_(stdout, cpu);
}


// memory size [MB]
static void memory_size(int *cpu, int *gpu)
{
	int64_t mem = 0;

	mem += 6 * (int64_t)NN * sizeof(real_t);        // Ex, Ey, Ez, Hx, Hy, Hz
	mem += 6 * (int64_t)NN * sizeof(id_t);          // iEx, iEy, iEz, iHx, iHy, iHz

	if (VECTOR) {
		mem += 12 * (int64_t)NN * sizeof(real_t);   // K1Ex, K2Ex, ... K1Hx, K2Hx, ...
	}

	// PML
	if (iABC == 1) {
		mem += (int64_t)(numPmlEx + numPmlEy + numPmlEz
		               + numPmlHx + numPmlHy + numPmlHz)
		     * (sizeof(pml_t) + 2 * sizeof(real_t));
	}

	// cEx, cEy, cEz, cHx, cHy, cHz
	const int64_t cpu_mem = mem + (int64_t)NFreq2 * Nx * Ny * Nz * 12 * sizeof(float);
	const int64_t gpu_mem = mem + (int64_t)NFreq2 * NN * 12 * sizeof(float);

	*cpu = (int)(cpu_mem / 1024 / 1024) + 1;
	*gpu = (int)(gpu_mem / 1024 / 1024) + 1;
}


// output filesize [MB]
static int output_size(void)
{
	const int64_t l = (iABC == 1) ? cPML.l : 1;
	const int64_t nsurf = 2 * ((Nx * Ny) + (Ny * Nz) + (Nz * Nx));
	const int64_t n3d = NFreq2 * (Nx + 2 * l) * (Ny + 2 * l) * (Nz + 2 * l) * 12 * sizeof(float);
	const int64_t s2d = NFreq2 * nsurf * (6 * sizeof(d_complex_t) + sizeof(surface_t));

	return (int)((n3d + s2d) / 1024 / 1024) + 1;
}
